llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5755,8 +5745,12 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
   // the InterleaveCount as if vscale is '1', although if some information about
   // the vector is known (e.g. min vector size), we can make a better decision.
   if (BestKnownTC) {
-    MaxInterleaveCount =
-        std::min(*BestKnownTC / VF.getKnownMinValue(), MaxInterleaveCount);
+    if (InterleaveSmallLoopScalarReduction)
+      MaxInterleaveCount =
+          std::min(*BestKnownTC / VF.getKnownMinValue(), MaxInterleaveCount);
+    else
+      MaxInterleaveCount = std::min(*BestKnownTC / (VF.getKnownMinValue() * 2),
```

Comment:
It feels like there are two independent changes in this patch that don't need to be tied together. Logically it looks like you could split these up into two patches:

1. For the general case (i.e. trip counts >= TinyTripCountInterleaveThreshold) change the MaxInterleaveCount so that we execute at least two vector loop iterations. Followed by ...
2. A patch to remove TinyTripCountInterleaveThreshold and add the InterleaveSmallLoopScalarReduction case here.

If you can split these up into two patches it would be easier for reviewers to see what test changes correspond to which code change. At the moment it's quite a large patch! What do you think?

Reply:
Sounds like a great idea!

Reply:
Working on this.

Reply:
Added part 1 of the changes for interleaving count in https://github.com/llvm/llvm-project/pull/70141

Reply:
Created 2 separate PRs on which this PR depends. 
1. https://github.com/llvm/llvm-project/pull/70272 - pre-committed additional tests
2. https://github.com/llvm/llvm-project/pull/70141 - part 1 of the patch that changes MaxInterleaveCount for a subset of cases. Added an additional condition of keeping MaxInterleaveCount same as before when the trip count is divisible by the VF, since that would generally mean that the epilogue loop is not required. 

Note that in the case of PGO-driven trip count, the trip count maybe an average over different TCs & not an exact one, allowing an epilogue loop to run even then - this case is not handled in this patch.